modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/RexUtils.java

```diff
@@ -328,6 +328,88 @@ public static IndexConditions buildSortedIndexConditions(
         return new IndexConditions(lower, upper, lowerBound, upperBound);
     }
 
+    /**
+     * Builds index conditions.
+     */
+    public static IndexConditions buildHashIndexConditions(
+            RelOptCluster cluster,
+            List<String> indexedColumns,
+            RexNode condition,
+            RelDataType rowType,
+            ImmutableBitSet requiredColumns
+    ) {
+        if (condition == null) {
+            return new IndexConditions();
+        }
+
+        condition = RexUtil.toCnf(builder(cluster), condition);
+
+        Int2ObjectOpenHashMap<List<RexCall>> fieldsToPredicates = mapPredicatesToFields(condition, cluster);
+
+        if (nullOrEmpty(fieldsToPredicates)) {
+            return new IndexConditions();
+        }
+
+        List<RexNode> searchCondition = new ArrayList<>();
+
+        Mappings.TargetMapping toTrimmedRowMapping = null;
+        if (requiredColumns != null) {
+            toTrimmedRowMapping = Commons.mapping(requiredColumns, rowType.getFieldCount());
+        }
+
+        for (int i = 0; i < indexedColumns.size(); i++) {
+            String columnName = indexedColumns.get(i);
+
+            RelDataTypeField field = rowType.getField(columnName, true, false);
+
+            if (field == null) {
```

Comment:
which case you cover by the condition? I can't imaging situation when field will be absent